lib/charms/opensearch/v0/opensearch_peer_clusters.py

```diff
@@ -0,0 +1,374 @@
+                self._charm.status.clear(message)
+            return
+
+        self._charm.app.status = BlockedStatus(deployment_desc.state.message)
+
+    def clear_directive(self, directive: Directive):
+        """Remove directive after having applied it."""
+        deployment_desc = self.deployment_desc()
+        if not deployment_desc:
+            return
+
+        if directive not in deployment_desc.directives:
+            return
+
+        deployment_desc.directives.remove(directive)
+        self._charm.peers_data.put_object(
+            Scope.APP, "deployment-description", deployment_desc.to_dict()
+        )
+
+    def deployment_desc(self) -> Optional[DeploymentDescription]:
+        """Return the deployment description object if any."""
+        current_deployment_desc = self._charm.peers_data.get_object(
+            Scope.APP, "deployment-description"
+        )
+        if not current_deployment_desc:
+            return None
+
+        return DeploymentDescription.from_dict(current_deployment_desc)
+
+    def validate_roles(self, nodes: List[Node], on_new_unit: bool = False) -> None:
+        """Validate full-cluster wide the quorum for CM/voting_only nodes on services start."""
+        deployment_desc = self.deployment_desc()
+        if not set(deployment_desc.config.roles) & {"cluster_manager", "voting_only"}:
+            # the user is not adding any cm nor voting_only roles to the nodes
+            return
+
+        if deployment_desc.start == StartMode.WITH_GENERATED_ROLES:
+            # the roles are automatically generated, we trust the correctness
+            return
+
+        # validate the full-cluster wide count of cm+voting_only nodes to keep the quorum
+        current_cluster_planned_units = self._charm.app.planned_units()
+        current_cluster_units = [
+            unit.name.replace("/", "-")
+            for unit in self._charm.model.get_relation(PeerRelationName).units
+        ]
+        current_cluster_online_nodes = [
+            node for node in nodes if node.name in current_cluster_units
+        ]
+
+        if len(current_cluster_online_nodes) < current_cluster_planned_units - 1:
+            # this is not the latest unit to be brought online, we can continue
+            return
+
+        voters = sum(1 for node in nodes if node.is_cm_eligible() or node.is_voting_only())
+        if voters % 2 == (0 if on_new_unit else 1):
+            # if validation called on new unit: it means it will start and maintain the quorum
+            #    (called on the latest unit to be configured and brought online)
+            # if validation called on existing cluster we should expect an odd number in the sum
+            return
+
+        raise OpenSearchProvidedRolesException(PClusterWrongNodesCountForQuorum)
+
+    def is_peer_cluster_relation_set(self):
+        """Return whether the peer cluster relation is established."""
+        return PeerClusterRelationName in self._charm.model.relations
```

Comment:
Maybe I would also factor this to a different PR, when this relation will be implemeted. If one wants to still keep this function, one can probably `return False`, with a comment that this will be implemented once the relation exists

Reply:
same reply as [here](https://github.com/canonical/opensearch-operator/pull/133#discussion_r1366721892)